packages/utils/src/lib/file-system.ts

```diff
@@ -87,3 +88,42 @@ export async function importEsmModule<T = unknown>(
   }
   return parse(mod.default);
 }
+
+export async function crawlFileSystem<T = string>(options: {
+  directory: string;
+  pattern?: string | RegExp;
+  fileTransform?: (filePath: string) => Promise<T> | T;
+}): Promise<T[]> {
+  const {
+    directory,
+    pattern,
+    fileTransform = (filePath: string) => filePath as T,
+  } = options;
+
+  const files = await readdir(directory);
+  const promises = files.map(async (file): Promise<T | T[]> => {
+    const filePath = join(directory, file);
+    const stats = await stat(filePath);
+
+    if (stats.isDirectory()) {
+      return crawlFileSystem({ directory: filePath, pattern, fileTransform });
+    } else if (stats.isFile() && (!pattern || new RegExp(pattern).test(file))) {
+      return fileTransform(filePath);
+    } else {
+      return [];
+    }
```

Comment:
Bit of a nitpick, but when you `return` in the `if`-block you don't need the `else` branch - helps flatten the code a bit.

```suggestion
    if (stats.isDirectory()) {
      return crawlFileSystem({ directory: filePath, pattern, fileTransform });
    }
    if (stats.isFile() && (!pattern || new RegExp(pattern).test(file))) {
      return fileTransform(filePath);
    }
    return [];
```